### PR TITLE
Support passing custom filters with the same name as built-in flags

### LINKED ISSUE #140: inconsistent behavior of filter pulling from context based on name
**Migrated issue, originally created by Michael Bayer ([@zzzeek](https://github.com/zzzeek))**

this is related to #3 but seems different

    from mako.template import Template
    
    x = Template("""
    X:
        ${"asdf" | h.foo}
    """)
    
    y = Template("""
    Y:
        ${"asdf" | z.foo}
    """)
    
    z = Template("""
    Z:
        ${h}
        ${"asdf" | h.foo}
    """)
    
    class h(object):
        foo = str
    
    try:
        print x.render(h=h)
    except NameError, e:
        print e
    
    print y.render(z=h)
    print z.render(h=h)

### PR DESCRIPTION
**Pull request by cocolato**
Fix: https://github.com/sqlalchemy/mako/issues/140

During the lexical analysis phase, add an additional prefix for undeclared identifiers that have the same name as built-in flags, and determine the final filter to be used during the code generation phase based on the context provided by the user.

```python3
from mako.template import Template

x = Template("""
X:
    ${"asdf" | h.foo}
""")

class h(object):
    foo = str

try:
    print x.render(h=h)
except NameError, e:
    print e
```
This code can now be rendered correctly.

### CODE REVIEW

File: mako/codegen.py [hunk 1]
```diff
@@ -522,6 +523,8 @@ def write_variable_declares(self, identifiers, toplevel=False, limit=None):
             self.printer.writeline("loop = __M_loop = runtime.LoopStack()")
 
         for ident in to_write:
+            if ident.startswith(CONFLICT_PREFIX):
```
Comment (zzzeek) [inline mako/codegen.py]: this test ahead had me curious so I ran some benches:

```py
import timeit
import re

PREFIX = "__SOME_PREFIX_"

def test_and_replace(text):
    if text.startswith(PREFIX):
        text = text.replace(PREFIX, "")
    return text

def just_replace(text):
    text = text.replace(PREFIX, "")
    return text

PREFIX_REG = re.compile(r"^__SOME_PREFIX")

def re_replace(text):
    text = PREFIX_REG.sub(text, "")
    return text


for fn in ("test_and_replace", "just_replace", "re_replace"):
    for negpos in ("negative_text", "__SOME_PREFIX_positive_text"):
        result = timeit.timeit(
            f"{fn}('{negpos}')", f"from __main__ import {fn}"
        )

        print(f"{fn:18} {negpos:30}: {result}")

```

results show just calling .replace alone as best

```
test_and_replace   negative_text                 : 0.06235152017325163
test_and_replace   __SOME_PREFIX_positive_text   : 0.11323748202994466
just_replace       negative_text                 : 0.037486436776816845
just_replace       __SOME_PREFIX_positive_text   : 0.06953103607520461
re_replace         negative_text                 : 0.08705236483365297
re_replace         __SOME_PREFIX_positive_text   : 0.08718583732843399

```

File: mako/codegen.py [hunk 2]
```diff
@@ -785,16 +788,36 @@ def locate_encode(name):
             else:
                 return filters.DEFAULT_ESCAPES.get(name, name)
 
-        if "n" not in args:
+        filter_args = []
+        conflict_n = "%sn" % CONFLICT_PREFIX
```
Comment (zzzeek) [inline mako/codegen.py]: might as well make this once at the top and just call it CONFLICT_PREFIX_N

or given my change to the variable name, DEFAULT_ESCAPED_N

File: mako/pyparser.py [hunk 3]
```diff
@@ -196,9 +198,20 @@ def visit_Tuple(self, node):
...
+            conflict_identifiers = undeclared_identifiers.intersection(
+                DEFAULT_ESCAPES
             )
+            if conflict_identifiers:
+                _map = {i: CONFLICT_PREFIX + i for i in conflict_identifiers}
+                for i, arg in enumerate(self.listener.args):
+                    if arg in _map:
+                        self.listener.args[i] = _map[arg]
+                self.listener.undeclared_identifiers = (
+                    undeclared_identifiers ^ conflict_identifiers
```
Comment (zzzeek) [inline mako/pyparser.py]: just for readability can we use the `.symmetric_difference` function here, since it's also using `union` and not `|`

File: test/test_ast.py [hunk 4]
```diff
@@ -285,16 +285,22 @@ def test_python_fragment(self):
...
-        parsed = ast.ArgumentList("h", **exception_kwargs)
-        eq_(parsed.args, ["h"])
+        parsed = ast.ArgumentList("m", **exception_kwargs)
+        eq_(parsed.args, ["m"])
+
+    def test_conflict_argument_list(self):
+        parsed = ast.ArgumentList(
+            "3, 5, 'hi', n+5, " "context.get('lala')", **exception_kwargs
+        )
+        eq_(parsed.undeclared_identifiers, {"__ALIAS_n", "context"})
```
Comment (zzzeek) [inline test/test_ast.py]: im a little confused by the focus on 'n', so does this test apply to other elements in DEFAULT_ESCAPES also like 'x', 'trim', etc?  I'd want to add tests for a bunch of those

File: mako/filters.py [hunk 5]
```diff
@@ -161,3 +161,5 @@ def htmlentityreplace_errors(ex):
     "str": "str",
     "n": "n",
 }
+
+CONFLICT_PREFIX = "__ALIAS_"
```
Comment (zzzeek) [inline mako/filters.py]: I think this would be better as

    DEFAULT_ESCAPE_PREFIX = "__DEFAULT_ESCAPE_"

File: mako/codegen.py [hunk 6]
```diff
@@ -785,16 +788,36 @@ def locate_encode(name):
...
+        if conflict_n not in args:
             if is_expression:
                 if self.compiler.pagetag:
                     args = self.compiler.pagetag.filter_args.args + args
-                if self.compiler.default_filters and "n" not in args:
+                    filter_args = self.compiler.pagetag.filter_args.args
```
Comment (zzzeek) [inline mako/codegen.py]: we're using `thing in filter_args` below so make this a set

File: mako/codegen.py [hunk 7]
```diff
@@ -785,16 +788,36 @@ def locate_encode(name):
...
+                if self.compiler.default_filters and conflict_n not in args:
                     args = self.compiler.default_filters + args
         for e in args:
-            # if filter given as a function, get just the identifier portion
-            if e == "n":
+            if e == conflict_n:
                 continue
+            if e.startswith(CONFLICT_PREFIX):
```
Comment (zzzeek) [inline mako/codegen.py]: this code is hard to follow because it's entirely separate from the code that was already here, even though it's doing mostly the same thing.    I think the new logic should be integrated with the existing logic rather than using an if: statement around the whole thing.    

If we start by setting up `e` into the two states that are needed:

```
if e.startswith(CONFLICT_PREFIX):
    render_e = e.replace(CONFLICT_PREFIX, "")
    is_default_filter = True
else:
    render_e = e
    is_default_filter = False
```


then the rest of the code can be continued without a big break, using is_default_filter as needed, this lets us see which parts of what we're doing are common to both

I did a step by step refactor, and came up with this, can you review?    

```py
        for e in args:
            if e == conflict_n:
                continue

            if e.startswith(CONFLICT_PREFIX):
                render_e = e.replace(CONFLICT_PREFIX, "")
                is_default_filter = e not in filter_args
            else:
                render_e = e
                is_default_filter = False

            # if filter given as a function, get just the identifier portion
            m = re.match(r"(.+?)(\(.*\))", e)

            if m:
                if not is_default_filter:
                    ident, fargs = m.group(1, 2)
                    f = locate_encode(ident)
                    render_e = f + fargs
                target = "%s(%s)" % (render_e, target)

            elif is_default_filter:
                target = "%s(%s) if %s is not UNDEFINED else %s(%s)" % (
                    render_e,
                    target,
                    render_e,
                    locate_encode(render_e),
                    target,
                )
            else:
                render_e = locate_encode(render_e)
                assert render_e is not None
                target = "%s(%s)" % (render_e, target)

        return target
```

**Review by zzzeek — CHANGES_REQUESTED** [review]
hey there!

lots of comments.  mostly looking to clean up variable names and integrate better in the filter logic, can you review a refactoring I'm proposing?   I did it step by step but I'm not 100% sure i didnt miss something, tests pass for each refactor step.

**Comment by cocolato** [comment]
> hey there!
> 
> 
> 
> lots of comments.  mostly looking to clean up variable names and integrate better in the filter logic, can you review a refactoring I'm proposing?   I did it step by step but I'm not 100% sure i didnt miss something, tests pass for each refactor step.


Thank you for taking the time and giving me so many suggestions and comments! I'm on a short trip and don't have a computer with me at the moment. I should be back home in two days to revise the current submission again🙏🙏

**Comment by cocolato** [comment]
The latest code has been updated. Previously, it might have been because I was concerned that making too many alterations to the code would impact other functions. Thus, I attempted to refrain from modifying the original logic and merely added some patch codes to it.

**Comment by cocolato** [comment]
I continued to test the efficiency issue regarding text replacement. The added "move_prefix" in Python 3.9 might be a better choice, but Mako currently only minimally supports Python 3.8.
```python3
import itertools
import re

from faker import Faker 

PREFIX = "__SOME_PREFIX_"
PREFIX_REG = re.compile(r"^__SOME_PREFIX")

class MyFaker(Faker):
    def prefix_name(self):
        return f"{PREFIX}{self.name()}"

fake = MyFaker()
data_num = 10000
escape_percent = 0.5
escape_num = int(data_num * escape_percent)

prefix_texts = [fake.prefix_name() for _ in range(escape_num)]
normal_texts = [fake.name() for _ in range(data_num - escape_num)]

def test_and_replace():
    for text in itertools.chain(prefix_texts, normal_texts):
        if text.startswith(PREFIX):
            res = text.replace(PREFIX, "")
    return res

def just_replace():
    for text in itertools.chain(prefix_texts, normal_texts):
        res = text.replace(PREFIX, "")
    return res

def move_prefix():
    for text in itertools.chain(prefix_texts, normal_texts):
        res = text.removeprefix(PREFIX)
    return res

def re_replace():
    for text in itertools.chain(prefix_texts, normal_texts):
        res = PREFIX_REG.sub(text, "")
    return res

__benchmarks__ = [
    (test_and_replace, just_replace, "just_replace"),
    (test_and_replace, move_prefix, "move_prefix"),
    (test_and_replace, re_replace, "re_replace")
]
```

bench result:
```
                                  Benchmarks, repeat=5, number=20                                   
┏━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃    Benchmark ┃ Min     ┃ Max     ┃ Mean    ┃ Min (+)         ┃ Max (+)         ┃ Mean (+)        ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│ just_replace │ 0.036   │ 0.038   │ 0.037   │ 0.017 (2.1x)    │ 0.018 (2.1x)    │ 0.017 (2.1x)    │
│  move_prefix │ 0.036   │ 0.038   │ 0.037   │ 0.010 (3.5x)    │ 0.011 (3.6x)    │ 0.010 (3.5x)    │
│   re_replace │ 0.036   │ 0.038   │ 0.037   │ 0.030 (1.2x)    │ 0.031 (1.2x)    │ 0.031 (1.2x)    │
└──────────────┴─────────┴─────────┴─────────┴─────────────────┴─────────────────┴─────────────────┘
```

**Comment by cocolato** [comment]
I continued to test the efficiency issue regarding text replacement. The added `remove_prefix` method in Python 3.9 might be a better choice, but Mako currently only minimally supports Python 3.8.

```python3
import itertools
import re

from faker import Faker 

PREFIX = "__SOME_PREFIX_"
PREFIX_REG = re.compile(r"^__SOME_PREFIX")

class MyFaker(Faker):
    def prefix_name(self):
        return f"{PREFIX}{self.name()}"

fake = MyFaker()
data_num = 10000
escape_percent = 0.5
escape_num = int(data_num * escape_percent)

prefix_texts = [fake.prefix_name() for _ in range(escape_num)]
normal_texts = [fake.name() for _ in range(data_num - escape_num)]

def test_and_replace():
    for text in itertools.chain(prefix_texts, normal_texts):
        if text.startswith(PREFIX):
            res = text.replace(PREFIX, "")
    return res

def just_replace():
    for text in itertools.chain(prefix_texts, normal_texts):
        res = text.replace(PREFIX, "")
    return res

def remove_prefix():
    for text in itertools.chain(prefix_texts, normal_texts):
        res = text.removeprefix(PREFIX)
    return res

def re_replace():
    for text in itertools.chain(prefix_texts, normal_texts):
        res = PREFIX_REG.sub(text, "")
    return res

__benchmarks__ = [
    (test_and_replace, just_replace, "just_replace"),
    (test_and_replace, remove_prefix, "remove_prefix"),
    (test_and_replace, re_replace, "re_replace")
]
```

benchmark res:
```
                                   Benchmarks, repeat=5, number=20                                   
┏━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┓
┃     Benchmark ┃ Min     ┃ Max     ┃ Mean    ┃ Min (+)         ┃ Max (+)         ┃ Mean (+)        ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━┩
│  just_replace │ 0.038   │ 0.043   │ 0.040   │ 0.018 (2.2x)    │ 0.018 (2.4x)    │ 0.018 (2.2x)    │
│ remove_prefix │ 0.038   │ 0.039   │ 0.039   │ 0.011 (3.5x)    │ 0.012 (3.3x)    │ 0.011 (3.5x)    │
│    re_replace │ 0.041   │ 0.048   │ 0.045   │ 0.034 (1.2x)    │ 0.039 (1.2x)    │ 0.037 (1.2x)    │
└───────────────┴─────────┴─────────┴─────────┴─────────────────┴─────────────────┴─────────────────┘
```

**Comment by zzzeek** [comment]
mainly I wanted to show that we can just have a single line with `replace()` and not do the extra check.  the goal is mostly less code

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 5e64e05c9370f82528654dcacee0d162889ce94a of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 5e64e05c9370f82528654dcacee0d162889ce94a: https://gerrit.sqlalchemy.org/c/sqlalchemy/mako/+/5557

**Comment by sqla-tester** [comment]
**Michael Bayer** (zzzeek) wrote:

great job

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/mako/+/5557

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/mako/+/5557 has been **merged**. Congratulations! :)
